cli-commands/verifier/verifier.js

```diff
@@ -0,0 +1,169 @@
+		this.libraries = this._buildLibrariesArguments(libraries);
+		this.apiUrl = await this._buildApiUrl(contractWrapper.provider);
+		this.solcVersionCompiler = this._buildSolcVersionCompiler(contractWrapper._contract.compiler.version);
+		logger.log(`Attempting to verify your contract: ${colors.colorName(this.contractName)} on network ${colors.colorParams(this.name)}`);
+		let data = this._constructRequestData();
+
+		const response = await this._sendVerificationRequest(data, defaultOverrides);
+		return await this._checkVerificationStatus(response, defaultOverrides);
+	}
+
+	async _flattenSourceCode(contractWrapper) {
+		const regexp = /[0-9]?\.[0-9]?\.[0-9]?/;
+		const solcVersion = regexp.exec(contractWrapper._contract.compiler.version)[0];
+		const sourceCode = await runWithoutWriteFiles(`${this.contractName}.sol`, solcVersion);
+		return sourceCode
+	}
+
+	async _buildConstructorArguments(contractWrapper, deploymentArguments) {
+		if (!deploymentArguments.length) {
+			return;
+		}
+		let encoder = new ethers.utils.AbiCoder();
+		let types = [];
+		contractWrapper.interface.deployFunction.inputs.forEach(element => {
+			types.push(element.type);
+		});
+		let encodedConstructorArgs = (await encoder.encode(types, deploymentArguments)).substr(2);
+		return encodedConstructorArgs;
+	}
+
+	_buildLibrariesArguments(libraries) {
+		if (!libraries || Object.keys(libraries).length === 0) {
+			return
+		}
+		let arrayLib = Object.entries(libraries).map(([key, value]) => ({ name: key, address: value }));
+		return arrayLib
+	}
+
+	_buildSolcVersionCompiler(solcVersion) {
+		const version = `v${solcVersion.replace(/.Emscripten.clang/g, '')}`;
+		return version;
+	}
+
+	async _buildApiUrl(provider) {
+		const { name } = await provider.getNetwork();
+		this.name = name;
+
+		if ((/^(ropsten|rinkeby|kovan|goerli)$/.test(this.name))) {
+			return `https://api-${this.name}.etherscan.io/api`;
+		}
+		return 'https://api.etherscan.io/api';
+	}
+
+	_constructRequestData() {
+		let data = {
+			apikey: this.apiKey,
+			module: 'contract', // DO NOT CHANGE
+			action: 'verifysourcecode', // DO NOT CHANGE
+			contractaddress: this.contractAddress,
+			sourceCode: this.flattenCode,
+			contractname: this.contractName,
+			compilerversion: this.solcVersionCompiler,
+			optimizationUsed: this.optimization,
+			runs: this.runs,
+			constructorArguements: this.constructorArguments
+		}
+		if (this.libraries) {
+			for (let i = 0; i < this.libraries.length; i++) {
+				data[`libraryname${i + 1}`] = this.libraries[i].name;
+				data[`libraryaddress${i + 1}`] = this.libraries[i].address;
+			}
+		}
+		let stringData = querystring.stringify(data);
+		return stringData
+	}
+
+	async _sendVerificationRequest(stringData, defaultOverrides) {
+
+		const ms = defaultOverrides.waitInterval || DEFAULT_SEND_REQUEST_TIMEOUT;
+		let count = 0;
+		const self = this;
+		async function sendRequest(ms) {
+			await self.timeout(ms)
```

Comment:
You are doing the timeout before the request. It is much better to leave the timeout for *after* the request